projects/com.oracle.truffle.llvm.asm.amd64/src/com/oracle/truffle/llvm/asm/amd64/AsmFactory.java

```diff
@@ -170,14 +183,44 @@ public void createDivisionOperatoin(String op, String divisor) {
             FrameSlot edxSlot = frameDescriptor.findFrameSlot("%edx");
             LLVMI32Node edxNode = LLVMI32ReadNodeGen.create(edxSlot);
             LLVMI32Node divisionNode = LLVMAMD64IdivlNodeGen.create(divisorNode, eaxNode, edxNode);
-            statement = LLVMWriteI32NodeGen.create(divisionNode, eaxSlot);
-            returnNode = LLVMI32ReadNodeGen.create(eaxSlot);
+            FrameSlot quotSlotEax = frameDescriptor.addFrameSlot("quotient");
+            quotSlotEax.setKind(FrameSlotKind.Int);
+            statements.add(LLVMWriteI32NodeGen.create(divisionNode, quotSlotEax));
+
+            if (retType == LLVMBaseType.STRUCT) {
+                LLVMI32Node remainderNode = LLVMAMD64ModNodeGen.create(divisorNode, eaxNode, edxNode);
```

Comment:
I think that you could compact this code by reading some information from the bitcode file, or by avoiding to create some of the frame slots. However, I think it is okay to do a refactoring the next time you implement a similar instruction, to help you identify common patterns.